src/core/libmaven/mzMassSlicer.cpp

```diff
@@ -109,99 +115,184 @@ void MassSlices::algorithmB( MassCutoff *massCutoff,int rtStep ) {
-                    s->ionCount = scan->intensity[k];
-                    s->rt=scan->rt;
-                    s->mz=mz;
-                    //cerr << "New Slice " <<  s->mzmin << " " << s->mzmax << " " << s->ionCount << endl;
-                    slices.push_back(s);
-                    int mzRange = mz * 10;
-                    cache.insert( pair<int,mzSlice*>(mzRange, s));
+                float intensity = scan->intensity[k];
+
+                // Checking if mz, intensity are within specified ranges
+                if (_maxMz && !isBetweenInclusive(mz,
+                                                  _minMz,
+                                                  _maxMz)) {
+                    continue;
+                }
+                if (_maxIntensity && !isBetweenInclusive(intensity,
+                                                         _minIntensity,
+                                                         _maxIntensity)) {
+                    continue;
                 }
+
+                // create new slice with the given bounds
+                float cutoff = massCutoff->massCutoffValue(mz);
+                mzSlice* s = new mzSlice(mz - cutoff,
+                                         mz + cutoff,
+                                         rt - rtWindow,
+                                         rt + rtWindow);
+                s->ionCount = intensity;
+                s->rt = scan->rt;
+                s->mz = mz;
+                slices.push_back(s);
             }
+
             // progress update 
             if (mavenParameters->showProgressFlag ) {
-                string progressText = to_string(i+1) + num + " out of " + to_string(mavenParameters->samples.size()) 
-                                  + " Sample(s) Processing.....\n"
-                                  + to_string(slices.size()) + " Slices Created ";
+                string progressText = to_string(i + 1)
+                                      + num
+                                      + " out of "
+                                      + to_string(mavenParameters->samples.size())
+                                      + " sample(s) processing…\n"
+                                      + to_string(slices.size())
+                                      + " slices created ";
                 sendSignal(progressText,currentScans,totalScans);
             }
+        }
+    }
 
+    cerr << "Found " << slices.size() << " slices" << endl;
+
+    // before reduction sort by mz first then by rt
+    sort(begin(slices),
+         end(slices),
+         [](const mzSlice* slice, const mzSlice* compSlice) {
+             if (slice->mz == compSlice->mz) {
+                 return slice->rt < compSlice->rt;
+             }
+             return slice->mz < compSlice->mz;
+         });
+
+    // reduction: merge slices sharing significant ROI
+    for (auto first = begin(slices); first != end(slices); ++first) {
```

Comment:
Would suggest creating a separate method for getting the vector of all slices. Just for cleanliness' sake.

Reply:
Alright. This one function has gotten too large anyway.